File: packages/sdk/typescript/src/writeback-consumer.ts
import type { RelayFileClient } from "./client.js";
import type { ConnectionProvider } from "./connection.js";
import type { WritebackItem } from "./types.js";

export interface WritebackHandler {
  canHandle(path: string): boolean;
  execute(item: WritebackItem, provider: ConnectionProvider): Promise<void>;
}

export interface WritebackConsumerOptions {
  client: RelayFileClient;
  workspaceId: string;
  handlers: WritebackHandler[];
  provider: ConnectionProvider;
  pollIntervalMs?: number;
  signal?: AbortSignal;
}

const DEFAULT_POLL_INTERVAL_MS = 1_000;

export class WritebackConsumer {
  private readonly client: RelayFileClient;
  private readonly workspaceId: string;
  private readonly handlers: WritebackHandler[];
  private readonly provider: ConnectionProvider;
  private readonly pollIntervalMs: number;
  private readonly signal?: AbortSignal;
  private loopPromise?: Promise<void>;
  private stopped = false;

  constructor(opts: WritebackConsumerOptions) {
    this.client = opts.client;
    this.workspaceId = opts.workspaceId;
    this.handlers = opts.handlers;
    this.provider = opts.provider;
    this.pollIntervalMs = opts.pollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS;
    this.signal = opts.signal;

    if (this.pollIntervalMs < 0) {
      throw new RangeError("pollIntervalMs must be greater than or equal to 0");
    }
  }

  async start(): Promise<void> {
    if (this.loopPromise) {
      if (this.stopped) {
        // Previous run was stopped — reset and start fresh
        this.stopped = false;
        this.loopPromise = undefined;
      } else {
        return this.loopPromise;
      }
    }
    this.loopPromise = this.runLoop();
    return this.loopPromise;
  }

  stop(): void {
    this.stopped = true;
  }

  async pollOnce(signal = this.signal): Promise<void> {
    const items = await this.client.listPendingWritebacks(this.workspaceId, undefined, signal);

    for (const item of items) {
      if (signal?.aborted || this.stopped) {
        return;
      }

      const handler = this.handlers.find((candidate) => candidate.canHandle(item.path));

      if (!handler) {
        await this.ackFailure(item, new Error(`No writeback handler found for path: ${item.path}`), signal);
        continue;
      }

      let executed = false;
      try {
        await handler.execute(item, this.provider);
        executed = true;
      } catch (error) {
        if (isAbortError(error) || signal?.aborted) {
          throw error;
        }
        await this.ackFailure(item, error, signal);
      }

      if (executed) {
        try {
          await this.client.ackWriteback({
            workspaceId: this.workspaceId,
            itemId: item.id,
            success: true,
            correlationId: item.correlationId,
            signal
          });
        } catch (ackError) {
          if (isAbortError(ackError) || signal?.aborted) {
            throw ackError;
          }
          // Handler succeeded — do NOT send failure ack (would cause duplicate writes on retry).
          // Log and move on; server will eventually time out and retry the ack.
        }
      }
    }
  }

  private async runLoop(): Promise<void> {
    while (!this.stopped && !this.signal?.aborted) {
      try {
        await this.pollOnce(this.signal);

        if (this.stopped || this.signal?.aborted) {
          return;
        }

        await sleep(this.pollIntervalMs, this.signal);
      } catch (error) {
        if (isAbortError(error) || this.signal?.aborted || this.stopped) {
          return;
        }
        throw error;
      }
    }
  }

  private async ackFailure(item: WritebackItem, error: unknown, signal?: AbortSignal): Promise<void> {
    await this.client.ackWriteback({
      workspaceId: this.workspaceId,
      itemId: item.id,
      success: false,
      error: toErrorMessage(error),
      correlationId: item.correlationId,
      signal
    });
  }
}

async function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  if (ms === 0) {
    return;
  }

  if (signal?.aborted) {
    throw signal.reason ?? new DOMException("The operation was aborted", "AbortError");
  }

  await new Promise<void>((resolve, reject) => {
    const onAbort = () => {
      clearTimeout(timer);
      reject(signal!.reason ?? new DOMException("The operation was aborted", "AbortError"));
    };

    const timer = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve();
    }, ms);

    if (!signal) {
      return;
    }

    signal.addEventListener("abort", onAbort, { once: true });
  });
}

function isAbortError(error: unknown): boolean {
  return error instanceof DOMException && error.name === "AbortError";
}

function toErrorMessage(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }

  return String(error);
}

File: packages/sdk/python/src/relayfile/writeback_consumer.py
from __future__ import annotations

import time
from typing import Protocol

from .client import RelayFileClient
from .connection import ConnectionProvider
from .types import AckWritebackInput, WritebackItem


class WritebackHandler(Protocol):
    def can_handle(self, path: str) -> bool:
        ...

    def execute(self, item: WritebackItem, provider: ConnectionProvider) -> None:
        ...


class WritebackConsumer:
    def __init__(
        self,
        *,
        client: RelayFileClient,
        workspace_id: str,
        handlers: list[WritebackHandler],
        provider: ConnectionProvider,
        poll_interval_ms: int = 1_000,
        ack_max_attempts: int = 3,
    ) -> None:
        if poll_interval_ms < 0:
            raise ValueError("poll_interval_ms must be greater than or equal to 0")
        if ack_max_attempts < 1:
            raise ValueError("ack_max_attempts must be greater than or equal to 1")
        self._client = client
        self._workspace_id = workspace_id
        self._handlers = handlers
        self._provider = provider
        self._poll_interval_ms = poll_interval_ms
        self._ack_max_attempts = ack_max_attempts
        self._stopped = False

    def start(self) -> None:
        self._stopped = False
        while not self._stopped:
            self.poll_once()
            if self._stopped:
                return
            time.sleep(self._poll_interval_ms / 1000)

    def stop(self) -> None:
        self._stopped = True

    def poll_once(self) -> None:
        items = self._client.list_pending_writebacks(self._workspace_id)
        for item in items:
            if self._stopped:
                return
            handler = next(
                (candidate for candidate in self._handlers if candidate.can_handle(item.path)),
                None,
            )
            if handler is None:
                self._ack_failure(item, Exception(f"No writeback handler found for path: {item.path}"))
                continue
            try:
                handler.execute(item, self._provider)
            except Exception as exc:
                self._ack_failure(item, exc)
                continue
            self._ack_with_retry(
                AckWritebackInput(
                    workspace_id=self._workspace_id,
                    item_id=item.id,
                    success=True,
                    correlation_id=item.correlation_id,
                )
            )

    def _ack_failure(self, item: WritebackItem, error: Exception) -> None:
        self._ack_with_retry(
            AckWritebackInput(
                workspace_id=self._workspace_id,
                item_id=item.id,
                success=False,
                error=str(error),
                correlation_id=item.correlation_id,
            )
        )

    def _ack_with_retry(self, ack: AckWritebackInput) -> None:
        # A handler side effect may already be committed before the ACK, so a
        # transient ACK failure must not redeliver the item. Retry with bounded
        # backoff; handlers should still be idempotent keyed by ``item.id``.
        last_error: Exception | None = None
        for attempt in range(self._ack_max_attempts):
            try:
                self._client.ack_writeback(ack)
                return
            except Exception as exc:  # noqa: BLE001 - surfaced after retries exhaust
                last_error = exc
                if attempt < self._ack_max_attempts - 1:
                    time.sleep(min(0.25 * (2**attempt), 2.0))
        raise RuntimeError(
            f"ack_writeback failed after {self._ack_max_attempts} attempts"
        ) from last_error
